src/core/dev_api/openvino/core/paged_cache_manager.hpp

```diff
@@ -0,0 +1,252 @@
+    void evict_one_unlocked();
+
+    std::vector<std::size_t> acquire_blocks_unlocked(size_t node_id, std::size_t block_count);
+    void copy_blocks_into_buffers_unlocked(const void* key_src_bytes,
+                                           const void* value_src_bytes,
+                                           const std::vector<std::size_t>& block_idxs);
+    void set_scores_for_blocks_unlocked(size_t node_id,
+                                        const std::vector<std::size_t>& block_idxs,
+                                        const float* scores);
+
+    static float cast_score_to_float(ov::element::Type et, const void* src_scalar) noexcept;
+    static bool is_element_compatible_with_T(ov::element::Type et, size_t sizeofT) noexcept;
+
+    void rebuild_evict_heap_unlocked();
+    static bool heap_less(const std::vector<block_t>& blocks, std::size_t a, std::size_t b) noexcept;
+
+private:
+    static void* aligned_allocate(std::size_t size);
+    static void aligned_free(void* p);
+
+    const ov::element::Type m_elem_type{};
+    const std::size_t m_total_bytes{0};
+
+    std::size_t m_block_size{0};
+    std::size_t m_block_bytes{0};  // m_block_size * elem_type_size
+    std::size_t m_num_heads{0};
+    std::size_t m_key_head_size{0};
+    std::size_t m_value_head_size{0};
+
+    std::size_t m_num_blocks{0};
+
+    void* m_key_base{nullptr};
+    void* m_value_base{nullptr};
+
+    std::vector<block_t> m_blocks;
+    std::list<std::size_t> m_free_block_list;
+    std::unordered_map<size_t, operator_state> m_ops;
+
+    std::vector<std::size_t> m_evict_heap;
+
+    mutable std::mutex m_mutex;
+};
+
+// -------- inline trivials & templates --------
+
+inline void* PagedCacheManager::get_key_base() const noexcept {
+    return get_cache_blocks().key_base;
+}
+inline void* PagedCacheManager::get_value_base() const noexcept {
+    return get_cache_blocks().value_base;
+}
+inline std::size_t PagedCacheManager::get_total_bytes() const noexcept {
+    return m_total_bytes;
+}
+inline ov::element::Type PagedCacheManager::get_element_type() const noexcept {
+    return m_elem_type;
+}
+inline std::size_t PagedCacheManager::get_num_blocks() noexcept {
+    return m_num_blocks;
+}
+inline std::size_t PagedCacheManager::get_block_size() noexcept {
+    return m_block_size;
+}
+inline std::size_t PagedCacheManager::get_block_bytes() noexcept {
+    return m_block_bytes;
+}
+
+template <typename T>
+std::vector<std::size_t> PagedCacheManager::insert(size_t node_id,
+                                                   const T* key_src,
+                                                   const T* value_src,
+                                                   std::size_t block_count,
+                                                   const T* scores) {
+    if (!is_element_compatible_with_T(m_elem_type, sizeof(T))) {
+        throw std::runtime_error("PagedCacheManager::insert<T>: T does not match element type");
```

Comment:
Don't use throw, Use openvino assertion instead.
Is exception required at all maybe data are validate during model validation process?